src/webgpu/api/operation/resource_init/buffer.spec.ts

```diff
@@ -5,7 +5,276 @@ export const description = `
 Test uninitialized buffers are initialized to zero when read
 (or read-written, e.g. with depth write or atomics).
 
-TODO
+Here lists all the buffer usages to test:
+- in writeBuffer()
+- in map async (read and partial write)
+- created with mappedAtCreation === true
+- as copy source (TODO)
+- as copy destination in a partial copy (TODO)
+- in ResolveQuerySet() (TODO)
+- as uniform / read-only storage / storage buffer (TODO)
+- as vertex / index buffe (TODO)
+- as indirect buffer (TODO)
 `;
 
 export const g = makeTestGroup(GPUTest);
+
+g.test('partial_write_buffer')
+  .paramsSubcasesOnly(u => u.combine('offset', [0, 8, -12]))
+  .fn(async t => {
+    const { offset } = t.params;
+    const bufferSize = 32;
+    const appliedOffset = offset >= 0 ? offset : bufferSize + offset;
+
+    const buffer = t.device.createBuffer({
+      size: bufferSize,
+      usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST,
+    });
+
+    const copySize = 12;
+    const writeData = new Uint8Array(copySize);
+    const expectedData = new Uint8Array(bufferSize);
+    for (let i = 0; i < copySize; ++i) {
+      expectedData[appliedOffset + i] = writeData[i] = i + 1;
+    }
+    t.queue.writeBuffer(buffer, appliedOffset, writeData, 0);
+
+    t.expectContents(buffer, expectedData);
+  });
+
+g.test('map_read_whole_buffer').fn(async t => {
+  const bufferSize = 32;
+
+  const buffer = t.device.createBuffer({
+    size: bufferSize,
+    usage: GPUBufferUsage.COPY_DST | GPUBufferUsage.MAP_READ,
+  });
+
+  await buffer.mapAsync(GPUMapMode.READ);
+  const readData = new Uint8Array(buffer.getMappedRange());
+  for (let i = 0; i < readData.length; ++i) {
+    t.expect(readData[i] === 0);
+  }
+  buffer.unmap();
+});
+
+g.test('map_read_partial_buffer')
+  .paramsSubcasesOnly(u => u.combine('offset', [0, 8, -16]))
+  .fn(async t => {
+    const { offset } = t.params;
+    const bufferSize = 32;
+    const appliedOffset = offset >= 0 ? offset : bufferSize + offset;
+
+    const buffer = t.device.createBuffer({
+      size: bufferSize,
+      usage: GPUBufferUsage.COPY_DST | GPUBufferUsage.MAP_READ,
+    });
+
+    {
+      const mapSize = 16;
+      await buffer.mapAsync(GPUMapMode.READ, appliedOffset, mapSize);
+      const readData = new Uint8Array(buffer.getMappedRange(appliedOffset, mapSize));
+      for (let i = 0; i < mapSize; ++i) {
+        t.expect(readData[i] === 0);
+      }
+      buffer.unmap();
+    }
+
+    {
```

Comment:
nit: I think this is the same as `map_read_whole_buffer`

Ah I realized it is because we can't do expectContents. Maybe add a small comment here and in other similar places below.